Transport/Measure-EmailDelayInMTL.ps1

```diff
@@ -0,0 +1,172 @@
+﻿# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+<#
+.NOTES
+	Name: Measure-EmailDelayInMTL.ps1
+	Requires: User Rights
+    Major Release History:
+        08/05/2024 - Initial Release
+
+.SYNOPSIS
+Generates a report of the maximum message delay for all messages in an Message Tracking Log.
+
+.DESCRIPTION
+Gather message tracking log details of all message to / from a given recipient for a given time range.
+Recommend using Start-HistoricalSearch in EXO.
+
+The script will provide an output of all unique message ids with the following information:
+MessageID
+Time Sent
+Total Time in transit
+
+Useful for determining if a "slow" message was a one off or a pattern.
+
+.PARAMETER MTLFile
+MTL File to process.
+
+.PARAMETER ReportPath
+Folder path for the output file.
+
+
+.OUTPUTS
+CSV File with the following information.
+    MessageID               ID of the Message
+    TimeSent                First time we see the message in the MTL
+    TimeReceived            Last delivery time in the MTL
+    MessageDelay            How long before the message was delivered
+
+Default Output File:
+$PSScriptRoot\MTL_report.csv
+
+.EXAMPLE
+.\Measure-EmailDelayInMTL -MTLPath C:\temp\MyMtl.csv
+
+Generates a report from the MyMtl.csv file.
+
+#>
+
+[CmdletBinding()]
+param (
+    [Parameter()]
+    [string]
+    $MTLFile,
+    [Parameter()]
+    [string]
+    $ReportPath = $PSScriptRoot
+)
+
+. $PSScriptRoot\..\Shared\ScriptUpdateFunctions\Test-ScriptVersion.ps1
+
+Function Test-CSVData {
+    param(
+        [array]$CSV,
+        [array]$ColumnsToCheck
+    )
+
+    $ColumnHeaders = ($CSV | Get-Member -MemberType NoteProperty).Name
+    ForEach ( $ColumnToCheck in $ColumnsToCheck) {
+        If ($ColumnHeaders.Contains($ColumnToCheck) ) {
+            # Nothing to do.
+        } Else {
+            Return $false
+        }
+    }
+    Return $true
+}
+
+if (Test-ScriptVersion -AutoUpdate) {
+    # Update was downloaded, so stop here.
+    Write-Host "Script was updated. Please rerun the command."
+    return
+}
+
+# make sure out output variable is null
+$output = $Null
+
+# Test for the provided file and load it.
+if (Test-Path $MTLFile) {
+    Write-Output "Loading MTL file."
+    $mtl = Import-Csv $MTLFile -Encoding Unicode
+} else {
+    Write-Error "Unable to find specified file"
+}
+
+# Validate the MTL
+if (Test-CSVData -CSV $mtl -ColumnsToCheck "event_id", "source", "message_id", "date_time_utc") {}
```

Comment:
`"event_id", "source", "message_id", "date_time_utc"` are not typical values in a Message Tracking Log. Is this the expected output from a cmdlet? If so, we need to be more clear about this script/script name. If it isn't intended for detecting Email Delay in all types of Message Tracking Logs, including Exchange On Prem, then we need to make that clear with the name and the parameters being used.

Reply:
added onprem MTL detection.  Detects if the MTL is from an onprem PS command and updates the info in memory to work just like an EXO one.